### PR TITLE
Eliminate double-free in realloc'd memory

### PR DESCRIPTION
**Pull request by gsjaardema**
If `memio->memory` is freed, then there is a double-free error since that pointer was either already freed in the realloc call, or it is the same pointer as newmen and should not be freed.   The documentation of realloc(3) states:
```
If the area pointed to was moved, a free(ptr) is done.
```

I'm not sure why I haven't hit this before, but I had a couple of my tests start throwing `double-free` errors and valgrind confirmed that the pointer had already been freed at the point of the realloc call.

### CODE REVIEW

**Comment by edhartnett** [comment]
@gsjaardema I did the same thing, but disabled a different test.

What is failing for me is dap4_test/test_remote.sh (see issue #1220).

**Comment by gdsjaar** [comment]
That is probably the one I intended to disable also…
..Greg

**Comment by DennisHeimbigner** [comment]
Odd that I did not see this as part of
pr https://github.com/Unidata/netcdf-c/pull/1202

**Comment by DennisHeimbigner** [comment]
Can  I see one of your test cases that causes the double free?

**Comment by DennisHeimbigner** [comment]
The problem is that using -fsanitize with gcc, I see no such errors.
So, the problem is related to the program you are using.

**Comment by gsjaardema** [comment]
I see the following:
```
s1018191:build(master)> gcc -fsanitize=address -o test-realloc -g test-realloc.c
s1018191:build(master)> ./test-realloc
=================================================================
==98893==ERROR: AddressSanitizer: attempting double-free on 0x614000000040 in thread T0:
    #0 0x107ac310d in wrap_free (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x5710d)
    #1 0x10740aee2 in main test-realloc.c:10
    #2 0x7fff5ba98014 in start (libdyld.dylib:x86_64+0x1014)

0x614000000040 is located 0 bytes inside of 400-byte region [0x614000000040,0x6140000001d0)
freed by thread T0 here:
    #0 0x107ac32d7 in wrap_realloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x572d7)
    #1 0x10740aed2 in main test-realloc.c:8
    #2 0x7fff5ba98014 in start (libdyld.dylib:x86_64+0x1014)

previously allocated by thread T0 here:
    #0 0x107ac2f53 in wrap_malloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56f53)
    #1 0x10740aeb3 in main test-realloc.c:5
    #2 0x7fff5ba98014 in start (libdyld.dylib:x86_64+0x1014)

SUMMARY: AddressSanitizer: double-free (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x5710d) in wrap_free
==98893==ABORTING
```
Which shows the same issue that valgrind shows.  I also get the following even without sanitize:
```
s1018191:build(master)> gcc -o test-realloc -g test-realloc.c
s1018191:build(master)> ./test-realloc
test-realloc(123,0x7fff944de380) malloc: *** error for object 0x7ffd32c02950: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```
I know this isn't NetCDF code, but it (I think) follows the same execution path as the NetCDF memio code.  It also matches the documentation of realloc as best I can see -- the free is not allowed since the memory has already been freed by realloc.

**Comment by DennisHeimbigner** [comment]
Could this issue be related to this (from docs/inmemory.md)?
> 2. If the *NC_MEMIO_LOCKED* flag is <b>not</b> set, then
> the netcdf library will take control of the incoming memory.
> This means that the user should not make any attempt to free
> or even read the incoming memory block in this case.

**Comment by gsjaardema** [comment]
If NC_MEMIO_LOCKED is set, then the reallocate is not called so that isn’t the issue.  Also, the memory freeing is inside netcdf code and has nothing to do with the client doing anything to the memory.

**Comment by DennisHeimbigner** [comment]
To make sure I understand, you are saying that your code does not set NC_MEMIO_LOCKED?
Correct?

**Comment by gsjaardema** [comment]
Yes

**Comment by gsjaardema** [comment]
Note that a run of ctest with or without `-fsanitize` will not catch any errors since the code in question is not covered by any of the existing NetCDF tests.  I added an `abort()` right after the `realloc` call and the test suite still passes.   This is in `memio_pad_length`

There is another `realloc` call earlier in the file and that one is called ( in `memio_open`) since if I add an `abort()`, it will cause four tests to fail.  It is also of note that this `realloc` does *not* do the calls to `free` and there are no memory leaks.

So, in summary, I think the pull request (without the test disables) is valid and should be applied.  Based on valgrind, -fsanitize, and the man page for realloc, and the other use of realloc in the same file, the memory should *not* be freed.

**Comment by gsjaardema** [comment]
The reason that the code in question is not being tested is that the file that is being created for the in_memory test is smaller than the initial size of memory allocated in `memio_create`.  Due to matching the pagesize, the memory is allocated at 4096 bytes.  The size of the file is only 300 bytes, so the realloc is never called.  

If we make a larger file by changing `#define DIM1_LEN 800` in `nc_test/tst_inmemory.c`, then the resulting file will be larger than 4096 bytes and then `memio_pad_length` will be called and you should get a memory failure.

If the PR is then applied, the test will pass.
